mlton/backend/ssa2-to-rssa.fun

```diff
@@ -1642,38 +1642,235 @@ fun convert (program as S.Program.T {functions, globals, main, ...},
+                         pushKeep st)
+                     fun real r =
+                        (setGlobalStatic (var, SOME (Real r));
+                         pushKeep st)
+                     fun 'a static (mk: S.Type.t * location * 'a ->
+                                    PackedRepresentation.StaticOrElem.t option, isEmpty, a: 'a) =
+                        let
+                           val location = getLocation (ty, isEmpty)
+                           val so = mk (ty, location, a)
+                           val rty = toRtype ty
+                        in
+                           case (so, rty) of
+                               (SOME s, SOME rty) =>
+                                 (case (s, location) of
+                                      (* address of heap static not a valid elem *)
+                                      (Static s, Heap) =>
+                                       pushStatic (var, ty, rty, s)
+                                    | (Static s, _) =>
+                                       (setGlobalStatic (var, SOME (Address var));
+                                        pushStatic (var, ty, rty, s))
+                                    | (Elem e, _) =>
+                                       (setGlobalStatic (var, SOME e);
+                                        pushKeep st))
+                             | _ => keep ()
+                        end
+                  in
+                     case exp of
+                        S.Exp.Const c =>
+                           (case c of
+                               Const.WordVector wxv => static (makeWordXVector, false, wxv)
+                             | Const.Word w => word w
+                             | Const.Real r => real r
+                             | _ => keep ())
+                      | S.Exp.Inject {variant, ...} =>
+                           copy variant
+                      | S.Exp.Object {args, con} =>
+                           let
+                              val location = getLocation (ty, false)
+                           in
+                              if Vector.exists (args, Option.isNone o globalStatic)
+                              (* For objects, there's no reason to statically initalize,
+                               * since they're so small and irregular *)
+                              orelse location = Static.Location.Heap
+                                 then keep ()
+                              else static (makeObject, false, {args=args, con=con})
+                           end
+                      | S.Exp.PrimApp {args, prim, ...} =>
+                           (case Prim.name prim of
+                               Prim.Name.Array_alloc _ =>
+                               let
+                                  val length = (globalStatic o Vector.first) args
+                               in
+                                  case (length, !Control.staticAllocArrays) of
```

Comment:
If `!Control.staticAllocArrays` is `true`, but `!Control.staticAllocVectors` is `false`, then `getLocation` will return `Heap` and the array will be created via `objectInits` (which corresponds to dynamically allocated / statically initialized?).  Just noting that `Control.staticAllocVectors` does not exclusively control `WordXVector`s.

Reply:
Related to https://github.com/MLton/mlton/pull/328#discussion_r321280198.

Reply:
Oh, I see now where the issue is. validCon using staticAllocVectors regardless of mutability is a bug.

Reply:
Resolved by 4a49d4d and deb6081.